v1/format/format.go

```diff
@@ -437,9 +454,41 @@ func (w *writer) writeComments(comments []*ast.Comment) {
 }
 
 func (w *writer) writeRules(rules []*ast.Rule, comments []*ast.Comment) []*ast.Comment {
+	var err error
 	for i, rule := range rules {
-		comments = w.insertComments(comments, rule.Location)
-		comments = w.writeRule(rule, false, comments)
+		// backup comments in case there is an error writing a rule
+		backupComments := make([]*ast.Comment, len(comments))
+		for i, c := range comments {
+			v := *c
+			backupComments[i] = &v
+		}
+
+		comments, err = w.insertComments(comments, rule.Location)
+		if err != nil {
+			panic(err)
+		}
+
+		// backup the current buffer in case of an error
+		var backup bytes.Buffer
+		backup.Write(w.buf.Bytes())
+		comments, err = w.writeRule(rule, false, comments)
+		if err != nil {
```

Comment:
Is there a reason for doing error handling here, where we write the rule, or could we confine the fix to `writeTerm` or `writeObject`? That way, we might cover more edge cases. Or does that produce broken formatted Rego?

Reply:
great idea, moved to `writeTerm` doesn't seem like it is needed in `writeObject` or at least haven't found a test case for it yet.

Reply:
more tests added 👍